doc/configuration.md
The CLI (command line interface) is how the user can configure the ESC, and also perform some simple tests.

# Getting into the CLI

## Required: USB Linker

To install, update, and configure Hydra firmware, you must use a USB-linker meant for communicating with ESCs. These are used with AM32, BLHeli_32, and BLHeli_S, for configuration and firmware updating.

See this [wiki page](../../../wiki/USB-Linker) for a list of USB-linkers you can use. That page also has links to the appropriate USB drivers that you might need to install.

![](imgs/usblinkerconnection.jpg)

## Required: Serial Terminal App

You need a simple serial terminal app of some sort installed on your computer.

For Windows, some of the popular choices: [Tera Term](https://tera-term.en.softonic.com/), [PuTTY](https://putty.org/), [RealTerm](https://realterm.sourceforge.io/), Arduino IDE's built-in serial monitor, some of the extensions available in VSCode

For Linux, perhaps try: `Minicom`, `picocom`, `screen`, `gtkterm`

Please remember to choose the correct serial port (the name will vary), and then configure the serial port for 19200 baud rate, 8 bits, no parity.

![](imgs/teraterm-start.png)

![](imgs/teraterm-baud.png)

SparkFun Electronics has a [tutorial about serial terminals](https://learn.sparkfun.com/tutorials/terminal-basics/real-term-windows)

## Connecting the USB linker

These steps are different from installing the firmware. Please do not confuse the two different instructions.

 1. Make sure the ESC is disconnected from everything. Do not connect the USB linker to it just yet. Keep the power off.

 2. Power on the ESC. The LED should indicate that it is disarmed (LED will stay solid RED, or whatever colour is used).

 3. Wait 10 seconds. After 10 seconds have passed, the LED blinking should have changed to a different pattern ![](imgs/ledblinks/cliwait.png) (if there is no LED, then the ESC will make the motor start buzzing at a high pitch)

 4. Connect the USB linker to the computer

 5. Connect the USB linker to the ESC

 6. Run the serial terminal app

The LED blink pattern will change again to indicate successfully entering CLI mode ![](imgs/ledblinks/climode.png)

Send the enter keystroke into the terminal a few times and you should see `>` appear on your screen. This means you are ready to enter in a command.

![](imgs/terminalprompt.png)

NOTE: arrow keys will never work, the delete or backspace key will simply cancel the entire command

NOTE: the CLI will never exit until you reboot the entire ESC with the USB linker disconnected

## The most important settings to change

 * `vsplitmode` is defaulted to the weakest mode that is easy to use, you might want to try out one of the boost modes
 * `inputmode` is obviously important if you are using CRSF or had problems doing the input wiring
 * `phasemap` is important because you don't know which motor output phase is the default common-shared phase (see [this guide](finding-common-shared-phase.md))

# Using the CLI

## Reading current settings

The command is `read`

It will output a giant list of all of the settings.

![](imgs/listallsettings.png)

You can copy this block of text into a text file for safekeeping.

(there's a section later describing each of the setting items)

## Changing a setting

The command looks like `<setting-name> <value>`, for example, `inputmode 1`.

A confirmation will shown if you are successful.

If you pasted in multiple lines, it will change multiple settings all at once. So you may paste in the text file you saved earlier.

WARNING: please don't go crazy with the values. The CLI is only smart enough to know if you forgot to include a value, or if you made a spelling mistake, but it cannot actually check if your value is within the correct range.

NOTE: for settings that are either true or false, please use `1` or `0`

(there's a section later describing each of the setting items)

## Factory Reset

The command is `factoryreset`

This will change every setting back to their default values.

## Reboot

The command is `reboot`

This will cause the ESC to enter into the AM32 bootloader.

# Available Settings and What They Mean

| command          | default | range       | description |
|------------------|---------|-------------|-------------|
| `vsplitmode    ` | 0       | 0 - 2       | voltage splitting mode <br /> 0 = boosted almost always <br  /> 1 = half voltage <br /> 2 = boosted while not turning |
| `loadbal       ` | 0       | 0 or 1      | lowers load on common-shared MOSFET |
| `inputmode     ` | 0       | 0 - 3       | 0 = RC PWM pulse input <br /> 1 = CRSF input <br /> 2 = RC PWM input through SWCLK <br /> 3 = CRSF input through SWCLK |
| `tankmix       ` | 0       | 0 or 1      | enables arcade tank drive mixing |
| `phasemap      ` | 1       | 1 - 3       | selects which one of the three phases is used as the [common-shared phase](finding-common-shared-phase.md) |
| `baud          ` | 0       |             | baud rate used for CRSF <br /> 0 means default (420000) <br /> useful for taking commands from slower microcontrollers |
| `channel_1     ` | 1       | 0 - 16      | which CRSF channel is the 1st channel, 0 = unassigned |
| `channel_2     ` | 2       | 0 - 16      | which CRSF channel is the 2nd channel, 0 = unassigned |
| `channel_mode  ` | 0       | 0 - 16      | which CRSF channel can control the voltage splitting mode, useful for temporary voltage boosting, 0 = unassigned |
| `channel_brake ` | 0       | 0 - 16      | which CRSF channel can control the braking mode, 0 = unassigned |
| `rc_mid        ` | 1500    | 800 - 2200  | RC PWM pulse middle value, in microseconds, does not affect CRSF |
| `rc_range      ` | 500     | 0 - 1000    | RC PWM pulse range, 500 means 1000 to 2000, does not affect CRSF |
| `rc_deadzone   ` | 10      | 0 - uint16  | RC PWM pulse deadzone around the center <br /> used if the signal is noisy <br /> unit is in microseconds <br /> does also apply to CRSF |
| `pwm_period    ` | ????    | 0 - 65535   | PWM period, generator reload value, controls the PWM frequency <br /> unit is in clock ticks <br /> Do not modify <br /> lower value = higher frequency
| `pwm_deadtime  ` | 50      | 0 - uint32  | PWM cannot be always 100% since the MOSFET drivers use a charge pump. High value means lower power for the motors, but if the value is too low then the high-side MOSFETs might stop working at full throttle. |
| `braking       ` | 1       | 0 or 1      | If true, enables complementary PWM and never floats the output pins. If false, disables complementary PWM and floats the output pins if throttle is 0 |
| `chanswap      ` | 0       | 0 or 1      | swaps the output channels |
| `flip1         ` | 0       | 0 or 1      | flips (reverse) the 1st output channel |
| `flip2         ` | 0       | 0 or 1      | flips (reverse) the 2nd output channel |
| `tied          ` | 0       | 0 or 1      | one channel controls both motors |
| `armdur        ` | 100     | 0 - uint32  | arming duration required, the number of valid arming pulses (zero throttle) required before becoming armed <br /> setting 0 means **always armed** |
| `disarmtime    ` | 3000    | 0 - uint32  | number of milliseconds of signal loss before going into disarmed state <br /> setting 0 means **never** |
| `templim       ` | 0       | 0 - 255     | temperature limit in degrees-Celcius <br /> setting 0 means limit disabled |
| `currlim       ` | 0       | 0 - 255     | current limit in milliamps <br /> setting 0 means limit disabled |
| `voltlim       ` | 0       | 0 - 65535   | low battery voltage limit in millivolts <br /> setting 0 means limit disabled <br /> please read [this guide](current-limiting-and-low-battery-limit.md) to understand this item |
| `cellmaxvolt   ` | 0       | 0 - 65535   | per cell maximum voltage in millivolts <br /> setting 0 means "use absolute limit instead" <br /> please read [this guide](current-limiting-and-low-battery-limit.md) to understand this item |
| `lowbattstretch` | 0       | 0 - 65535   | determines if reaching the low-battery limit will stop the motors or simply slow it down <br /> please read [this guide](current-limiting-and-low-battery-limit.md) to understand this item |
| `voltdiv       ` | ????    | 0 - uint32  | hardware voltage divider adjustment <br /> default is dependant of hardware, do not adjust if you don't need to |
| `curroffset    ` | ????    | 0 - uint32  | hardware current sensor offset adjustment <br /> default is dependant of hardware, do not adjust if you don't need to |
| `currscale     ` | ????    | 0 - uint32  | hardware current sensor scaling adjustment <br /> default is dependant of hardware, do not adjust if you don't need to |
| `adcfilter     ` | 10      | 1 - 100     | analog signal low-pass filtering constant, for avoiding noisy sensor measurements <br /> low value means strong filtering, slow changes <br /> high value means weak filtering, fast changes <br /> 0 and 100 both mean "disable filter" |
| `curlimkp      ` | 100     | percentage  | PID constant P for the current limiting feature <br /> do not adjust, 100% is default and ideal |
| `curlimki      ` | 0       | sint32      | PID constant I for the current limiting feature <br /> do not adjust |
| `curlimkd      ` | 100     | percentage  | PID constant D for the current limiting feature <br /> do not adjust, 100% is default and ideal |

There are other items that are not always used, they are described in other usage guides.

## Voltage Boosting During Runtime

If you use the CRSF input mode, then you can configure a switch on your RC radio transmitter to change the voltage-splitting-mode live. The configuration item `vsplitmode` will not be used, and the switch will determine what voltage-splitting-mode you are operating in.

This means you can be in the very linear (easy to use) half-voltage mode most of the time, but hit a button and have extra pushing power for your robot.

To do this, simply configure the item `channel_mode` to the CRSF channel the switch is mapped to.

The switch channel output value maps to the voltage-split-mode according to this table

| switch value <br /> in microseconds | voltage split mode |
|-------------------------------------|--------------------|
| < 1333                              | boosted almost always |
| 1334 to 1666                        | half-voltage       |
| 1667 >                              | boosted while not turning |

# Advanced Commands

## Version

Use the command `version` to show some version information.

## Debugging Hardware

The command is `hwtest`, using it will enter into a test mode, it will ask you to press a key to choose a test mode.

 * key `v` will help you do voltage sensor calibration
 * key `c` will help you do current sensor calibration
 * key `p` will help you do test phase mapping

[Please see this sensor-calibration guide](sensor-calibration.md)

[Please see this hardware guide](other-hardware-hacking.md) for information on finding the common-shared phase


src-app/release_notes.txt
# V0.4 Release Notes

This is a alpha release, features are all complete, but not fully tested.

## New features


src-app/cli.cpp
#include "main.h"

// usage of the command-line-interface CLI is documented at https://github.com/frank26080115/Hydra-Dual-Brushed-Motor-ESC/blob/master/doc/configuration.md

#ifdef ENABLE_COMPILE_CLI

#include <stdlib.h>
#include <stdio.h>

#include "cereal.h"
#include "swd_pins.h"

#include "inputpin.h"
#include "led.h"
#include "phaseout.h"
#include "userconfig.h"
#include "sense.h"
#include "rc.h"
#include "tone.h"
#include "version.h"

extern uint16_t current_limit_duty;
extern void current_limit_task(int v1, int v2);

#if defined(DEVELOPMENT_BOARD)
extern Cereal_USART dbg_cer;
extern RcChannel* rc1;
extern RcChannel* rc2;
#endif

extern Cereal_USART        main_cer;
extern Cereal_TimerBitbang cli_cer;

void cli_reportSensors(Cereal*);
void cli_execute(Cereal* cer, char* str);
int cmd_parseArgs(char* str);
void eeprom_print_all(Cereal* cer);

extern "C" {
extern const EEPROM_item_t cfg_items[];
extern const EEPROM_data_t cfge;
extern int eeprom_quick_validate(void);
#ifndef RELEASE_BUILD
extern uint8_t eeprom_error_log;
#endif
}

void cli_enter(void)
{
    tone_start(4, 0, cfg.tone_volume);
    tone_setVolume(0);
    ledblink_cli();
    swdpins_deinit();
    dbg_printf("CLI entered at %u\r\n", millis());

    #if defined(MAIN_SIGNAL_PA2)
        #if defined(ARTERY)
            // Artery chips have no echo problem on the USART peripheral
            Cereal_USART* cer = &main_cer;
            cer->init(CEREAL_ID_USART2, CLI_BAUD, true, false);
        #else
            // STM32 chips have an echo problem when using half-duplex mode
            // https://github.com/frank26080115/Hydra-Dual-Brushed-Motor-ESC/issues/1
            // I have attempted to fix the problem but it is still not 100% reliable
            // so I've simply defaulted to using bit-bang cereal instead
            //Cereal_TimerBitbang* cer = &cli_cer;
            //cer->init(CLI_BAUD);

            // update: Feb 5 2024
            // hmmm it feels like it can work with USART
            Cereal_USART* cer = &main_cer;
            cer->init(CEREAL_ID_USART2, CLI_BAUD, true, false);
        #endif
    #elif defined(MAIN_SIGNAL_PB4)
    Cereal_TimerBitbang* cer = &cli_cer;
    cer->init(CLI_BAUD);
    #else
    Cereal_TimerBitbang* cer = &cli_cer;
    cer->init(CLI_BAUD);
    #endif

    char buff[CLI_BUFF_SIZE];
    uint16_t buff_idx = 0;
    char prev = '\0';
    bool has_interaction = false;
    bool prev_has_interaction = false;
    uint32_t last_idle_prompt = 0;

    while (true)
    {
        uint32_t tnow = millis();
        led_task(false);

        int16_t c = cer->read();
        if (c >= 0) // data available
        {
            eeprom_delay_dirty();

            has_interaction |= ((c >= 'a' || c >= 'A') && c != '>');

            if (prev_has_interaction == false && has_interaction) {
                // stop tone on first good keystroke
                prev_has_interaction = true;
                tone_stop();
            }

            if (c == '\n' && prev == '\r')
            {
                // ignore windows style new line sequence
                prev = c;
                buff_idx = 0;
                continue;
            }
            else if (c == '\r' || c == '\n' || c == '\0')
            {
                // commit end line
                prev = c;
                #if CLI_ECHO
                cer->write('\r');
                cer->write('\n');
                #endif
                buff[buff_idx] = '\0';
                if (buff_idx > 0) {
                    cli_execute(cer, buff);
                }
                buff_idx = 0;
                cer->write('\r');
                cer->write('\n');
                cer->write('>'); // show prompt
                continue;
            }
            else if (c == 0x08 || c == 0x1B || c == 0x18 || c == 0x7F)
            {
                // these keys will clear the buffer, cancelling anything that's been written
                prev = c;
                #if CLI_ECHO
                cer->write('\r');
                cer->write('\n');
                #endif
                buff_idx = 0;
                cer->write('\r');
                cer->write('\n');
                cer->write('>');
                continue;
            }
            else if ((c == ' ' || c == '\t' || c == '>') && buff_idx <= 0)
            {
                // prevent leading spaces
                prev = c;
                continue;
            }
            else if ((c == ' ' || c == '\t') && (prev == ' ' || prev == '\t'))
            {
                // prevent double spaces
                prev = c;
                continue;
            }
            else
            {
                prev = c;
                if (buff_idx < CLI_BUFF_SIZE - 2) {
                    // place in buffer if there is room
                    c = c == '\t' ? ' ' : c; // convert tabs to spaces
                    buff[buff_idx] = c;
                    buff_idx++;
                    //buff[buff_idx] = 0;
                    #if CLI_ECHO
                    cer->write(c);
                    #endif
                }
                continue;
            }
        }
        else if (has_interaction == false && (tnow - last_idle_prompt) >= 3000) {
            last_idle_prompt = tnow;
            // if the user has not interacted at all with the CLI
            // then keep printing the prompt symbol periodically
            // just to aid the user
            cer->write('>');
        }

        // do other things
        #if defined(RELEASE_BUILD)
        eeprom_save_if_needed();
        #else
        if (eeprom_save_if_needed()) {
            int i = eeprom_quick_validate();
            if (i != 0) {
                cer->printf("ERROR: EEPROM save validation failed %d\r\n", i);
            }
            else {
                cer->printf("\r\nEEPROM saved!\r\n");
            }
        }
        #endif
        sense_task();

        #if defined(DISABLE_LED) || defined(ENABLE_TONE)
        if (has_interaction == false)
        {
            if ((millis() % 700) < 200) {
                tone_setVolume(cfg.tone_volume);
            }
            else {
                tone_setVolume(0);
            }
        }
        #endif
    }
}

void cli_execute(Cereal* cer, char* str)
{
    //int argc;
    if (item_strcmp("read", str, NULL))
    {
        cer->printf("all settings:\r\n\r\n");
        eeprom_print_all(cer);
        cer->write('\r');
        cer->write('\n');
        #ifndef RELEASE_BUILD
        cer->printf("write count: %lu\r\n", cfg.write_cnt);
        #endif
    }
    else if (item_strcmp("version", str, NULL))
    {
        cer->printf("\r\nV%u.%u E %u HW 0x%08lX N:%s\r\n", firmware_info.version_major, firmware_info.version_minor, firmware_info.version_eeprom, firmware_info.device_code, firmware_info.device_name);
        #ifndef RELEASE_BUILD
        cer->printf("input pin is GPIO P%c%u\r\n"
            , (char)(((uint32_t)'A') + ((firmware_info.device_code & 0xFF) / ((((uint32_t)GPIOB_BASE) - ((uint32_t)GPIOA_BASE)) >> 8)))
            , (uint8_t)((firmware_info.device_code >> 8) & 0xFF)
        );
        cer->printf("CPU freq %lu\r\n", (uint32_t)SystemCoreClock);
        #endif
    }
    else if (item_strcmp("factoryreset", str, NULL))
    {
        eeprom_factory_reset();
        cer->printf("\r\nfactory reset done");
    }
    else if (item_strcmp("reboot", str, NULL))
    {
        cer->printf("\r\nrebooting...\r\n");
        cer->flush();
        NVIC_SystemReset();
    }
    else if (item_strcmp("hwtest", str, NULL))
    {
        pwm_set_remap(cfg.phase_map);
        pwm_set_period(PWM_DEFAULT_PERIOD);
        pwm_set_braking(true);
        pwm_set_all_duty_remapped(0, 0, 0);

        uint8_t  test_mode  = 0;
        uint16_t ori_vdiv   = cfg.voltage_divider;
        uint16_t ori_offset = cfg.current_offset;
        uint16_t ori_scale  = cfg.current_scale;
        int8_t   phase_map  = cfg.phase_map;
        uint32_t tick = 0;
        uint32_t tnow;
        uint16_t pwr = 0, pwr100 = 0;
        int16_t c;
        while (true)
        {
            led_task(false);
            sense_task();
            tnow = millis();
            c = cer->read();
            if (c == 'x' || c == 0x08 || c == 0x1B || c == 0x18 || c == 0x7F) {
                break;
            }
            if (test_mode == 'c')
            {
                switch (c)
                {
                    case ',' : cfg.current_offset   -=   1; break;
                    case '.' : cfg.current_offset   +=   1; break;
                    case '<' : cfg.current_offset   -=  10; break;
                    case '>' : cfg.current_offset   +=  10; break;
                    case ';' : cfg.current_scale    -=   1; break;
                    case '\'': cfg.current_scale    +=   1; break;
                    case ':' : cfg.current_scale    -=  10; break;
                    case '"' : cfg.current_scale    +=  10; break;
                }
            }
            else if (test_mode == 'v')
            {
                switch (c)
                {
                    case '[' : cfg.voltage_divider  -=   1; break;
                    case ']' : cfg.voltage_divider  +=   1; break;
                    case '{' : cfg.voltage_divider  -=  10; break;
                    case '}' : cfg.voltage_divider  +=  10; break;
                }
            }

            if ((test_mode == 'c' || test_mode == 'p') && c >= '0' && c <= '9') {
                pwr    = fi_map(c, '0', '9', 0, PWM_DEFAULT_PERIOD, true);
                pwr100 = fi_map(c, '0', '9', 0, 100               , true);
                pwm_set_all_duty_remapped(pwr, 0, 0);
            }
            if (test_mode == 'p' && (c == '-' || c == '=' || c == '+')) {
                pwm_set_all_duty_remapped(0, 0, 0);
                phase_map = (c == '-') ? (phase_map - 1) : (phase_map + 1);
                phase_map = (phase_map <= 0) ? 3 : phase_map;
                phase_map = (phase_map >  3) ? 1 : phase_map;
                pwm_set_remap(phase_map);
                pwm_set_all_duty_remapped(pwr, 0, 0);
            }

            if (c == 'v') {
                test_mode = c;
                pwr = 0;
                pwr100 = 0;
                pwm_set_all_duty_remapped(0, 0, 0);
            }
            else if (c == 'c' || c == 'p') {
                test_mode = c;
            }

            if ((tnow - tick) >= 250) {
                if (test_mode == 'v') {
                    cer->printf("[%lu] raw-v %u , calc-v %lu , v-div %lu\r\n", tnow, adc_raw_voltage, sense_voltage, cfg.voltage_divider);
                }
                else if (test_mode == 'c') {
                    cer->printf("[%lu] raw-c %u , calc-c %lu , ", tnow, adc_raw_current, sense_current); cer->flush();
                    cer->printf("c-offset %lu , c-scale %lu , pwr %u\r\n", cfg.current_offset, cfg.current_scale, pwr100);
                }
                else if (test_mode == 'p') {
                    cer->printf("[%lu] phase-map %u , pwr %u\r\n", tnow, phase_map, pwr100);
                }
                else {
                    cer->printf("[%lu] press 'v' or 'c' or 'p'\r\n", tnow);
                }
                tick = tnow;
            }
        }
        pwm_set_all_duty_remapped(0, 0, 0);

        cer->printf("\r\nsession end, voltdiv %lu , ", cfg.voltage_divider); cer->flush();
        cer->printf("curroffset %lu , currscale %lu , ", cfg.current_offset, cfg.current_scale); cer->flush();
        cer->printf("phasemap %d \r\n(not saved !!!)\r\n", phase_map);

        cfg.voltage_divider = ori_vdiv;
        cfg.current_offset  = ori_offset;
        cfg.current_scale   = ori_scale;
        pwm_set_remap(cfg.phase_map);
    }
    #ifndef RELEASE_BUILD
    else if (item_strcmp("viewlog", str, NULL))
    {
        cer->printf("\r\nEEPROM load log = %d\r\n", eeprom_error_log);
    }
    #endif
    else
    {
        // none of the commands match, loop through the potential settings to see if it's something to save
        int32_t nv;
        bool res = eeprom_user_edit(str, &nv);
        if (res) {
            cer->printf("\r\nOK (%s = %li)", str, nv);
            load_runtime_configs();
            eeprom_unlock_key = EEPROM_MAGIC;
            return;
        }
        else {
            // unrecognized command
            cer->printf("\r\nERROR CMD \"%s\"", str);
            return;
        }
    }
}

void eeprom_print_all(Cereal* cer)
{
    int i;
    for (i = 0; ; i++)
    {
        EEPROM_item_t* desc = (EEPROM_item_t*)&(cfg_items[i]);
        if (desc->ptr == 0) {
            return;
        }
        uint32_t ptrstart = (uint32_t)&cfge;
        uint32_t itmidx = desc->ptr - ptrstart;
        uint8_t* ptr8 = (uint8_t*)&cfg;
        int32_t v = 0;
        memcpy(&v, &(ptr8[itmidx]), desc->size);
        dbg_printf("%u %u \t ", itmidx, desc->size);
        cer->printf("%s %li\r\n", desc->name, v);
    }
}

extern uint8_t crsf_inputGuess;

#ifdef DEVELOPMENT_BOARD
void cliboot_if_key(void)
{
    if (dbg_cer.available() >= 3 && dbg_cer.peekTail() == '\n') {
        dbg_cer.reset_buffer();
        dbg_printf("CLI enter from keys\r\n");
        cli_enter();
    }
}
#else
#define cliboot_if_key(...)
#endif

void cliboot_decide(void)
{
    uint32_t tstart;
    inp_init();
    inp2_init();

    ledblink_cliwait();

    int i;
    for (i = 0; i < CLI_ENTER_UNPLUG_CRITERIA; i++)
    {
        inp_pullDown();
        inp2_pullDown();
        if (cfg.input_mode == INPUTMODE_RC_SWD || cfg.input_mode == INPUTMODE_CRSF_SWCLK) {
            swclk_init(LL_GPIO_PULL_DOWN);
        }

        tone_start(2, 150, cfg.tone_volume);

        tstart = millis();
        while ((millis() - tstart) < 3) {
            // wait for pull resistors to take effect
            if (inp_read() == 0
            #if defined(MAIN_SIGNAL_PB4)
            && ((cfg.input_mode == INPUTMODE_CRSF && inp2_read() == 0) || cfg.input_mode != INPUTMODE_CRSF)
            #endif
            && ((cfg.input_mode == INPUTMODE_CRSF_SWCLK && swclk_read() == 0) || cfg.input_mode != INPUTMODE_CRSF_SWCLK)
            )
            {
                break;
            }
        }

        bool is_1 = false, is_2 = false;
        while ((millis() - tstart) < 500)
        {
            led_task(false);
            cliboot_if_key();
            if ((is_2 |= (inp_read() != 0)) || (is_1 |= (inp2_read() != 0)) ||
                    ((cfg.input_mode == INPUTMODE_RC_SWD || cfg.input_mode == INPUTMODE_CRSF_SWCLK) && swclk_read() != 0))
            {
                dbg_printf("CLI cancel from signal high\r\n");
                if (is_1) {
                    crsf_inputGuess = 1;
                }
                else if (is_2) {
                    crsf_inputGuess = 2;
                }

                if (is_2)
                {
                    tstart = millis();
                    while (true)
                    {
                        led_task(false);
                        if (inp_read() == 0 || inp2_read() != 0 || ((cfg.input_mode == INPUTMODE_RC_SWD || cfg.input_mode == INPUTMODE_CRSF_SWCLK) && swclk_read() != 0)) {
                            // this will happen if a CRSF receiver is connected and transmissions start to happen
                            dbg_printf("reboot cancel from signal low\r\n");
                            return;
                        }
                        if ((millis() - tstart) >= 5000)
                        {
                            // if it reboots but exits the bootloader, it will probably end up at the `(inp_read() == 0)` check above
                            dbg_printf("reset into bootloader\r\n");
                            NVIC_SystemReset(); // to back to bootloader
                        }
                    }
                }

                return;
            }
        }
        tstart = millis();
        inp_pullUp();
        inp2_pullUp();
        if (cfg.input_mode == INPUTMODE_RC_SWD || cfg.input_mode == INPUTMODE_CRSF_SWCLK) {
            swclk_init(LL_GPIO_PULL_UP);
        }
        while ((millis() - tstart) < 3)
        {
            // wait for pull resistors to take effect
            if (inp_read() != 0
            #if defined(MAIN_SIGNAL_PB4)
            && ((cfg.input_mode == INPUTMODE_CRSF && inp2_read() != 0) || cfg.input_mode != INPUTMODE_CRSF)
            #endif
            && ((cfg.input_mode == INPUTMODE_CRSF_SWCLK && swclk_read() != 0) || cfg.input_mode != INPUTMODE_CRSF_SWCLK)
            )
            {
                break;
            }
        }
        while ((millis() - tstart) < 500)
        {
            led_task(false);
            cliboot_if_key();
            if (inp_read() == 0 || inp2_read() == 0 ||
                    ((cfg.input_mode == INPUTMODE_RC_SWD || cfg.input_mode == INPUTMODE_CRSF_SWCLK) && swclk_read() == 0))
            {
                dbg_printf("CLI cancel from signal low\r\n");
                return;
            }
        }
    }

    // pull down main signal input again and wait for it to settle
    inp_pullDown();
    tstart = millis();
    while ((millis() - tstart) < 3) {
        if (inp_read() == 0) {
            break;
        }
    }

    // wait for user to plug in
    ledblink_cliplug();
    tstart = millis();
    tone_start(2, 0, cfg.tone_volume);
    while (true)
    {
        led_task(false);

        #if defined(DISABLE_LED) || defined(ENABLE_TONE)
        uint32_t tspan = millis() - tstart;
        if ((tspan % 600) < 200) {
            tone_setVolume(cfg.tone_volume);
        }
        else {
            tone_setVolume(0);
        }
        #endif

        if (inp_read() != 0) {
            cli_enter();
        }
    }
}

#endif // ENABLE_COMPILE_CLI
